dimagi/commcare-hq
# Case search custom ES sort

## Product Description
Adds UI in app builder under _Case Search and Claim_ for _Custom Sort Properties_.
![image](https://github.com/dimagi/commcare-hq/assets/100609770/e6134cb2-2afa-4173-8a05-69578c4b3e71)

Through these properties, allows sorting of case search results as they are pulled from ElasticSearch, for example: sorting cases by date so that the most recent cases appear first in the search return, and if there were >500 cases, the cases over 500 that would be left out of the search return would be the older cases. It could be more accurate from a user's perspective to think of this as "priority" than "sorting" as it doesn't affect the order in which search results are displayed.

## Technical Summary
[USH-3673](https://dimagi-dev.atlassian.net/browse/USH-3673), [Spec](https://docs.google.com/document/d/1Cpvy4qN4Qy0YF7u8GT4BDzB8_kEOehOIIB-OrBy0Qwc/edit?pli=1)

Per spec, uses a `commcare_sort` key to add custom sort properties in a `<data>` element which is added to the suite. This is read back and parsed during a case search to apply sorting to the ES query in the order provided.

Can be reviewed commit-by-commit.

## Feature Flag
SYNC_SEARCH_CASE_CLAIM

## Safety Assurance

### Safety story
Tested locally with different properties, sort types and directions and reviewed results as returned from ElasticSearch for accurate sorting. Tested on staging with ~600 cases to see that results are returned differently in web apps based on custom sort properties (e.g., custom sort on `person` cases by age descending leaves out the youngest cases).

### Automated test coverage
Added automated tests for suite generation and ES query:
`corehq.apps.app_manager.tests.test_suite_remote_request.test_custom_sort_properties`
`corehq.apps.case_search.tests.test_case_search_es_queries.test_add_custom_sort_properties`


### QA Plan
QA should at least test that UI component works and saves correctly. Unsure about testing search results since this change is mostly invisible in web apps.

### Rollback instructions
If a case list had Custom Sort Properties added and this PR was reverted, case search would be broken on that case list until the next app build.

### Labels & Review


[USH-3673]: https://dimagi-dev.atlassian.net/browse/USH-3673?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
